web3/contract.py

```diff
@@ -89,6 +89,56 @@
 ACCEPTABLE_EMPTY_STRINGS = ["0x", b"0x", "", b""]
 
 
+class ContractFunctions(object):
+    """Class containing contract function objects
+    """
+
+    _method_names = []
+
+    def __init__(self, contract):
+        self.contract = contract
```

Comment:
I don't know if it's feasible but I'm curious if we can do this without passing in the whole contract.  In theory we *should* be able to get by with only passing the address, ABI and the web3 instance.  This should make this **significantly** easier to test in isolation.

Reply:
Maybe but given that it looks like a more complex change than I predicted I'll just open up an issue noting this as an area of minor debt that we can try to clean up at some point.  Skip it for now.

Reply:
Ive already removed the need to pass in the whole contract.  It was necessary to avoid a circular reference.  Ive left the helper functions as they are by copying some contract methods to the ContractMethod class.  I tried to make the required change minimal.  There is more clean up that can be done, so maybe that can be dealt with in a new issue.